File: day_02/go/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	inputPath := "./day_02/input.txt"
	fmt.Println("--- Part One ---")
	fmt.Println(part1(inputPath))

	fmt.Println("--- Part Two ---")
	fmt.Println(part2(inputPath))
}

func part1(inputPath string) int {
	course := readStrings(inputPath)
	hor := 0
	ver := 0
	for _, step := range course {
		dirValue := strings.Split(step, " ")
		direction := dirValue[0]
		value := toInt(dirValue[1])
		switch direction {
		case "forward":
			hor += value
		case "down":
			ver += value
		case "up":
			ver -= value
		}
	}
	return hor * ver
}

func part2(inputPath string) int {
	course := readStrings(inputPath)
	hor := 0
	ver := 0
	aim := 0
	for _, step := range course {
		dirValue := strings.Split(step, " ")
		direction := dirValue[0]
		value := toInt(dirValue[1])
		switch direction {
		case "forward":
			hor += value
			ver += aim * value
		case "down":
			aim += value
		case "up":
			aim -= value
		}
	}
	return hor * ver
}

func readStrings(filename string) []string {
	file, err := os.Open(filename)
	check(err)
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var text []string
	for scanner.Scan() {
		text = append(text, strings.TrimRight(scanner.Text(), "\n"))
	}
	return text
}

func check(err error) {
	if err != nil {
		panic(err)
	}
}

func toInt(s string) int {
	result, err := strconv.Atoi(s)
	check(err)
	return result
}

func max(numbers []int) int {
	currMax := numbers[0]
	for _, val := range numbers {
		if val > currMax {
			currMax = val
		}
	}
	return currMax
}

func min(numbers []int) int {
	currMin := numbers[0]
	for _, val := range numbers {
		if val < currMin {
			currMin = val
		}
	}
	return currMin
}

File: day_02/python/main.py
def part_one(filename: str) -> int:
    with open(filename) as f:
        course = map(lambda line: line.strip().split(), f.readlines())

    hor, ver = 0, 0
    for step in course:
        direction, amount = step[0], int(step[1])
        if direction == "forward":
            hor += amount
        elif direction == "down":
            ver += amount
        elif direction == "up":
            ver -= amount
    return hor * ver


def part_two(filename: str) -> int:
    with open(filename) as f:
        course = map(lambda line: line.strip().split(), f.readlines())

    hor, ver, aim = 0, 0, 0
    for step in course:
        direction, amount = step[0], int(step[1])
        if direction == "forward":
            hor += amount
            ver += aim * amount
        elif direction == "down":
            aim += amount
        elif direction == "up":
            aim -= amount
    return hor * ver


if __name__ == "__main__":
    input_path = "./day_02/input.txt"
    print("---Part One---")
    print(part_one(input_path))

    print("---Part Two---")
    print(part_two(input_path))
